Frameworks/CoreGraphics/DWriteFontCollectionHelper.h

```diff
@@ -0,0 +1,60 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <CoreGraphics/DWriteWrapper.h>
+
+#import <unordered_map>
+
+/**
+ * Struct that exposes CFHash and CFEqual to an STL container
+ */
+struct __CFStringHashEqual {
+    std::size_t operator()(const woc::unique_cf<CFStringRef>& item) const {
+        return CFHash(item.get());
+    }
+
+    bool operator()(const woc::unique_cf<CFStringRef>& item1, const woc::unique_cf<CFStringRef>& item2) const {
+        return CFEqual(item1.get(), item2.get());
+    }
+};
+
+using _DWriteFontPropertiesMap =
+    std::unordered_map<woc::unique_cf<CFStringRef>, std::shared_ptr<_DWriteFontProperties>, __CFStringHashEqual, __CFStringHashEqual>;
+
+/**
+ * A wrapper around an IDWriteFontCollection that provides mappings not directly available from the IDWriteFontCollection interface,
+ * such as font name -> family name, postscript name -> font name, font name -> font weight, style,
+ * by means of an internal cache.
+ *
+ * Concrete implementations must implement _GetFontCollection(), which is expected to return an IDWriteFontCollection to work on.
+ */
+class DWriteFontCollectionHelper {
+public:
+    DWriteFontCollectionHelper() {
+    }
+
+    CFMutableArrayRef CopyFontFamilyNames();
+    CFMutableArrayRef CopyFontNamesForFamilyName(CFStringRef familyName);
+    std::shared_ptr<_DWriteFontProperties> GetFontPropertiesFromUppercaseFontName(const woc::unique_cf<CFStringRef>& upperFontName);
+    HRESULT CreateFontFamilyWithName(_In_ const wchar_t* unicharFamilyName, _Out_ IDWriteFontFamily** outFontFamily);
```

Comment:
It fine to use SAL I guess (we don't get the benefit of actually doing anything useful with it except for reading clarity) but it seems to be inconsistent about when you are using it. Take the below _GetFontLIstForFamilyName method that doesn't annotate its in parameter. If we are going to do it, lets make it consistent so that COM style HRESULT returning functions all use SAL or whatever ...

---
In reply to: [90746813](https://github.com/Microsoft/WinObjC/pull/1504#discussion_r90746813) [](ancestors = 90746813,90743039,90738161)